### PR TITLE
gRPC Peer Selection

### PR DESCRIPTION
**Pull request by peter-edge**
This adds peer selection to gRPC. This comments out one part of config testing for now, I'd like to update config in a subsequent PR.

### CODE REVIEW

File: transport/x/grpc/integration_test.go [hunk 1]
```diff
@@ -159,6 +159,14 @@ func newTestEnv(inboundOptions []InboundOption, outboundOptions []OutboundOption
...
 
 	t := NewTransport()
+	if err := t.Start(); err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			err = multierr.Append(err, t.Stop())
+		}
+	}()
```
Comment (kriskowal) [inline transport/x/grpc/integration_test.go]: The nil check is not necessary for these. Append checks both sides for nil.

File: transport/x/grpc/outbound.go [hunk 2]
```diff
@@ -118,41 +134,32 @@ func (o *Outbound) invoke(
 	if responseMD != nil {
 		callOptions = []grpc.CallOption{grpc.Trailer(responseMD)}
 	}
+	apiPeer, onFinish, err := o.peerChooser.Choose(ctx, request)
+	defer func() {
+		if onFinish != nil {
```
Comment (kriskowal) [inline transport/x/grpc/outbound.go]: onFinish must not be nil. The peer lists are allowed to return a singleton, but the end of request needs to be tracked regardless.

File: transport/x/grpc/transport.go [hunk 3]
```diff
@@ -47,7 +57,18 @@ func (t *Transport) Start() error {
...
-	return t.once.Stop(nil)
+	return t.once.Stop(func() error {
+		t.lock.Lock()
+		defer t.lock.Unlock()
+		for _, grpcPeer := range t.addressToPeer {
+			grpcPeer.stop()
+		}
+		var err error
+		for _, grpcPeer := range t.addressToPeer {
+			err = multierr.Combine(err, grpcPeer.wait())
```
Comment (kriskowal) [inline transport/x/grpc/transport.go]: Can be Append since its duadic.

File: transport/x/grpc/options.go [hunk 4]
```diff
@@ -20,7 +20,11 @@
 
 package grpc
 
-import "google.golang.org/grpc"
+import (
+	"go.uber.org/yarpc/api/backoff"
```
Comment (peter-edge) [inline transport/x/grpc/options.go]: Yea I saw that - I prefer this, I appreciate you letting me keep it.

File: transport/x/grpc/peer.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+		clientConn: clientConn,
+		stoppingC:  make(chan struct{}, 1),
+		stoppedC:   make(chan error, 1),
+	}
+	go grpcPeer.monitor()
+	return grpcPeer, nil
+}
+
+func (p *grpcPeer) monitor() {
+	// wait for start so we can be certain that we have a channel
```
Comment (peter-edge) [inline transport/x/grpc/peer.go]: This actually verifies I call transport.Start() in testing, so I'd want to keep it anyways :-)

File: transport/x/grpc/integration_test.go (hunk 1 above)
Comment (peter-edge) [inline transport/x/grpc/integration_test.go]: In this case it is - I only want to call stop if there is an error being returned from the constructor

File: transport/x/grpc/outbound.go (hunk 2 above)
Comment (peter-edge) [inline transport/x/grpc/outbound.go]: This is more defensive programming

File: transport/x/grpc/transport.go (hunk 3 above)
Comment (peter-edge) [inline transport/x/grpc/transport.go]: Changed

**Pull request merged**